agent/unit/manager.go

```diff
@@ -0,0 +1,268 @@
+package unit
+
+import (
+	"sync"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/util/slice"
+)
+
+var (
+	ErrUnitNotFound             = "unit not found"
+	ErrUnitAlreadyRegistered    = "unit already registered"
+	ErrCannotUpdateOtherUnit    = "cannot update other unit's status"
+	ErrDependenciesNotSatisfied = "unit dependencies not satisfied"
+	ErrSameStatusAlreadySet     = "same status already set"
+	ErrCycleDetected            = "cycle detected"
+	ErrFailedToAddDependency    = "failed to add dependency"
+)
+
+// Status represents the status of a unit.
+type Status string
+
+// Status constants for dependency tracking
+const (
+	StatusNotRegistered Status = ""
+	StatusPending       Status = "pending"
+	StatusStarted       Status = "started"
+	StatusComplete      Status = "completed"
+)
+
+// ID provides a type narrowed representation of the unique identifier of a unit.
+type ID string
+
+// Unit represents a point-in-time snapshot of a vertex in the dependency graph.
+// Units may depend on other units, or be depended on by other units. The unit struct
+// is not aware of updates made to the dependency graph after it is initialized and should
+// not be cached.
+type Unit struct {
+	id     ID
+	status Status
+	// ready is true if all dependencies are satisfied.
+	// It does not have an accessor method on Unit, because a unit cannot know whether it is ready.
+	// Only the Manager can calculate whether a unit is ready based on knowledge of the dependency graph.
+	// To discourage use of an outdated readiness value, only the Manager should set and return this field.
+	ready bool
+}
+
+func (u *Unit) ID() ID {
+	return u.id
+}
+
+func (u *Unit) Status() Status {
+	return u.status
+}
+
+// Dependency represents a dependency relationship between units.
+type Dependency struct {
+	Unit           ID
+	DependsOn      ID
+	RequiredStatus Status
+	CurrentStatus  Status
+	IsSatisfied    bool
+}
+
+// Manager provides reactive dependency tracking over a Graph.
+// It manages Unit registration, dependency relationships, and status updates
+// with automatic recalculation of readiness when dependencies are satisfied.
+type Manager struct {
+	mu sync.RWMutex
+
+	// The underlying graph that stores dependency relationships
+	graph *Graph[Status, ID]
+
+	// Store vertex instances for each unit to ensure consistent references
+	units map[ID]Unit
+}
+
+// NewManager creates a new Manager instance.
+func NewManager() *Manager {
+	return &Manager{
+		graph: &Graph[Status, ID]{},
+		units: make(map[ID]Unit),
+	}
+}
+
+// Register adds a unit to the manager if it is not already registered.
+// If a Unit is already registered (per the ID field), it is not updated.
+func (m *Manager) Register(id ID) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.registered(id) {
+		return xerrors.Errorf("%s: unit %q", ErrUnitAlreadyRegistered, id)
```

Comment:
```suggestion
		return xerrors.Errorf("unit %q: %w", id, ErrUnitAlreadyRegistered)
```
I feel error wrapping would be more appropriate here. Allows error testing rather than string matching. Likewise the order above "local: wrapped" is more common in Go.